### PR TITLE
Add better Thread handling, reincorporate ConversationDelegate

### PR DESCRIPTION
**Pull request by lewisdaly**
The most significant change is that we setup app with a config object, to enable us to configure and inject our different dependencies throughout the application. This requires us to pass through app into every constructor (which I was avoiding, but now realise is the best way).

This is generally good practice, to avoid singletons, and allow us to inject our own test dependencies when we start writing tests (which is soon).

Also:
- Also added new envlocal.sh script, for settingup local environment.
  This requires users to define a STAGE variable (as demonstrated in the readme)
- added setup to disable redis
- added script for flushing redis

### CODE REVIEW

File: src/models/Thread.js [hunk 1]
```diff
@@ -161,13 +161,41 @@ class Thread {
...
+
+    //CompleteResponse contains the necessary user-facing message for us to ask for more info!
+    /*
+      Hmm this is a little unintutive. We set the state, call handle enter state, and THEN handle the response
+      for the previous state? I wonder how we could do it better.
+      The issue is that everything happens within a request lifecycle.
+
+      one option is to pass a promise chain of 'post response steps'
+      the response object will call the response, and look for any of these steps to perform after the response
+      we could define a custom object that could do this instead of just returning a response object as below
```
Comment (lewisdaly) [inline src/models/Thread.js:187]: Does this make sense?

File: src/index.js [hunk 2]
```diff
@@ -1,25 +1,34 @@
...
+const saveReadingDelegate = new ConversationDelegate(app, 'saveReading');
+conversationRouter.registerConversationDelegate(saveReadingDelegate);
 
+const botApi = new BotApi(app);
+app.set('config', {
+  botApi: botApi,
+  conversationRouter: conversationRouter,
+});
 
 app.get('/message', function (req, res) {
```
Comment (kevindoveton) [inline src/index.js:24]: should this be a post request?

**Comment by kevindoveton** [comment]
if redis is disabled what happens? do all required entities need to be in the message?

File: src/utils/RedisHelper.js [hunk 3]
```diff
@@ -1,11 +1,5 @@
 const isNullOrUndefined = require('util').isNullOrUndefined;
 const redis = require('redis');
-const bluebird = require('bluebird');
```
Comment (kevindoveton) [inline src/utils/RedisHelper.js:3]: bluebird dependency should be removed from package.json

**Comment by lewisdaly** [comment]
Disable redis is just a handy thing to have for unit testing, or just running things locally so we don't always get annoying state error messages.

File: src/utils/RedisHelper.js (hunk 3 above)
Comment (lewisdaly) [reply src/utils/RedisHelper.js:3]: Yep

File: src/index.js (hunk 2 above)
Comment (lewisdaly) [reply src/index.js:24]: No, this is a way to set some shared configuration across the application - just storing a dict. But yeah, it does like pretty similar to an actual request

Comment (kevindoveton) [reply src/index.js:24]: app.get('/message', function (req, res) {

i mean this

**Comment by lewisdaly** [comment]
Thanks for your comments Kev.

**Comment by lewisdaly** [comment]
Maybe, but I know that the Way2Mint API (SMS gateway for MyWell) sends a get and not post. 

So this is fine for now, eventually we will have another layer of abstraction in front of this anyway.

**Pull request merged**